src/mqtt/protocol/mqtt/mqtt_quic.c

```diff
@@ -1469,17 +1465,23 @@ mqtt_quic_sock_set_sqlite_option(
  ******************************************************************************/
 // allocate main stream with pipe
 static int
-quic_mqtt_stream_init(void *arg, nni_pipe *qsock, void *sock)
+quic_mqtt_stream_init(void *arg, nni_pipe *pipe, void *sock)
 {
 	bool         major = false;
 	mqtt_pipe_t *p     = arg;
-	p->qsock           = qsock;
 	p->mqtt_sock       = sock;
 	p->cparam          = NULL;
 
-	if (p->mqtt_sock->pipe == NULL) {
+	if (pipe != NULL) {
+		/* control stream */
 		p->mqtt_sock->pipe = p;
+		p->npipe = pipe
 		major = true;
+	} else {
+		/* data stream */
+		// rhack: dialer_start ? Initial the pipe to transport layer
+		//nni_dialer_start(d, 0);
+		//p->npipe = d->d_pipes? // How to get pipe in d_pipes
```

Comment:
nni_dialer_start should not be called here.
You should not call quic_mqtt_stream_init in protocol layer. On the contrary. it would be called by underly layer when a quic stream is established.